nvflare/widgets/fed_event.py

```diff
@@ -109,29 +116,61 @@ def _receive(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareab
         return make_reply(ReturnCode.OK)
 
     def _post(self):
-        sleep_time = 0.1
+        """
+        During ABOUT_TO_END_RUN, sleep_time is 0 and system will flush
+         in_events by firing events without delay.
+        During END_RUN, system will wait for self.grace_period, even the queue is empty,
+        so any new item can be processed.
+        However, since the system does not guarantee the receiving side of _post is still
+        alive, we catch the exception and show warning messages to users if events can not
+        be handled by receiving side.
+        """
+        sleep_time = self.regular_interval
+        countdown = self.grace_period
         while True:
             time.sleep(sleep_time)
-            if self.asked_to_stop or self.abort_signal.triggered:
+            if self.abort_signal.triggered:
                 break
-
-            with self.in_lock:
-                if len(self.in_events) <= 0:
+            n = len(self.in_events)
+            if n > 0:
+                if self.asked_to_flush:
+                    sleep_time = 0
+                else:
+                    sleep_time = self.regular_interval
+                with self.in_lock:
+                    event_to_post = self.in_events.pop(0)
+            elif self.asked_to_stop:
```

Comment:
if self.asked_to_stop is True,

should we stop adding events into self.in_events?

Reply:
After discussing with @yhwen , we decided not to block event receiving because components which hasn't received END_RUN might still firing fed events.  We even wait two seconds to collect those events and fire them during END_RUN to flush the queue.

Reply:
Got it, just a note that, for the "grace_period" we set.
We might wait a bit longer than that, because we only decrease the grace_period / countdown when length of in events is 0.